Build a regression model.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
# import linear_model from sklearn
from sklearn import linear_model

Provide model output and an interpretation of the results. 

In [3]:
analysis_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/analysis_df.csv', index_col=0)
analysis_df.head()


bar_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

   popularity  rating  price  review_count  num_stations_pinged  empty_slots  \
0        0.97     9.0    3.0          99.0                 11.0           12   
1        0.98     8.3    2.0          36.0                  9.0           12   
2        0.96     8.1    2.0         131.0                  8.0           12   
3        0.98     9.5    2.0          13.0                 23.0           12   
4        0.98     7.8    2.0         129.0                  6.0           12   

   free_bikes  total_slots  percent_used  
0           3           15          80.0  
1           3           15          80.0  
2           3           15          80.0  
3           3           15          80.0  
4           3           15          80.0

In [4]:
X = analysis_df[['popularity', 'rating', 'review_count', 'price', 'num_stations_pinged', 'free_bikes', 'total_slots']]
y = pd.Series(analysis_df['percent_used'],name=analysis_df['percent_used'][0])

In [5]:
X.shape

(487, 7)

In [6]:
X.head()

popularity  rating  review_count  price  num_stations_pinged  free_bikes  \
0        0.97     9.0          99.0    3.0                 11.0           3   
1        0.98     8.3          36.0    2.0                  9.0           3   
2        0.96     8.1         131.0    2.0                  8.0           3   
3        0.98     9.5          13.0    2.0                 23.0           3   
4        0.98     7.8         129.0    2.0                  6.0           3   

   total_slots  
0           15  
1           15  
2           15  
3           15  
4           15

In [7]:
y

0       80.0
1       80.0
2       80.0
3       80.0
4       80.0
       ...  
482     40.0
483     89.0
484    100.0
485    100.0
486    100.0
Name: 80.0, Length: 487, dtype: float64

#### Discussion of the Rating and number of nearby stations model output:

In [10]:
# Model for how the number of stations within 1000m affects the bar's rating
X = analysis_df[['popularity', 'rating', 'review_count', 'price', 'num_stations_pinged', 'free_bikes', 'total_slots']]
y = pd.Series(analysis_df['rating'],name=analysis_df['rating'][0])

In [11]:
X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X['num_stations_pinged'])

In [12]:
#Model for Number of Nearby Stations and Rating
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                                 OLS Regression Results                                
Dep. Variable:                    9.0   R-squared (uncentered):                   0.571
Model:                            OLS   Adj. R-squared (uncentered):              0.570
Method:                 Least Squares   F-statistic:                              646.0
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                    2.79e-91
Time:                        16:09:16   Log-Likelihood:                         -1484.9
No. Observations:                 487   AIC:                                      2972.
Df Residuals:                     486   BIC:                                      2976.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                          coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------

### Understanding the Model Outputs
In the model output above, you should pay attention to the following statistics in this order:

Model 1: The Rating and Percent Used:
R-squared: R-squared reflects the fit of the model. R-squared values range from 0 to 1, where a higher value generally indicates a better fit, assuming certain conditions are met. In this output, we can see that the value is 0.571. This means that the model is capable of explaining 57.1% of the patterns in the data. It does make sense that bars that have a more public transit options to get to them would have more of an opportunity for patrongage and therefore would have a higher rating. Also, more bike stations within 1000m suggests that the bar is more centrally located within the city and therefore would have more tourists and people looking for a good time which lends itself well to a higher rating. 

P>|t| (or the p-value): A p-value of less than 0.05 is considered to be statistically significant. This regression output shows a p-value of 0, which means that the probability of the relationship between the number of stations within 1000m of a bar and the bar's rating being solely due to natural variation is 0. In other words, the number of nearby bike stations does impact the rating of a bar.

coef: Coef is short for the coefficient of the independent variable. It represents the impact on the response variable per unit increase of the independent variable. In this case, we can see that the coef of num_statations_pinged has a value of 0.7047. This means that a unit increase in the value of num_stations_pinged will have a positive impact on the rating of a bar.

#### Single Linear Regression for how a bar's rating affects a bar's popularity:

In [15]:
# Model for the relationship of rating and popularity
X = analysis_df[['rating', 'review_count', 'price', 'num_stations_pinged', 'free_bikes', 'total_slots']]
y = pd.Series(analysis_df['popularity'],name=analysis_df['popularity'][0])

X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X['rating'])

#Model for Number of Nearby Stations and Rating
model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                                 OLS Regression Results                                
Dep. Variable:                   0.97   R-squared (uncentered):                   0.828
Model:                            OLS   Adj. R-squared (uncentered):              0.828
Method:                 Least Squares   F-statistic:                              2348.
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                   3.38e-188
Time:                        18:36:11   Log-Likelihood:                         -142.83
No. Observations:                 487   AIC:                                      287.7
Df Residuals:                     486   BIC:                                      291.9
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

I made a model for the relationship between popularity and rating because I wanted to be sure that they had strong positive relationship. I also think this is a strong relationship for bars to be able to know. Popularity is a deeply human and subjective criteria, but rating (which really is their online rating) is much more concrete. The model's p-value shows a statistically significant relationship between rating and popularity. The rating can account for 82.8% of the popularity rating of the bars considered in this model. Unsurprisingly, the positive coefficient means that as a bar's rating increases so too does the bar's popularity. 

#### Multi-linear Regression: With Popularity

In [17]:
# Model for Bar Ratings and Bike usage on nearby stations:
X = analysis_df[['popularity', 'review_count', 'price', 'num_stations_pinged', 'percent_used']]
y = pd.Series(analysis_df['rating'],name=analysis_df['rating'][0])
# Multiple Linear Regression:
X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X)


model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                    9.0   R-squared:                       0.081
Model:                            OLS   Adj. R-squared:                  0.071
Method:                 Least Squares   F-statistic:                     8.483
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           1.04e-07
Time:                        19:38:36   Log-Likelihood:                -685.94
No. Observations:                 487   AIC:                             1384.
Df Residuals:                     481   BIC:                             1409.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   7.3119    

In [18]:
# Model for Bar Ratings and Bike usage on nearby stations:
X = analysis_df[['review_count', 'price', 'num_stations_pinged', 'percent_used']]
y = pd.Series(analysis_df['rating'],name=analysis_df['rating'][0])
# Multiple Linear Regression:
X = sm.add_constant(X) # adding a constant
lin_reg = sm.OLS(y,X)


model = lin_reg.fit()
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                    9.0   R-squared:                       0.081
Model:                            OLS   Adj. R-squared:                  0.073
Method:                 Least Squares   F-statistic:                     10.61
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           3.01e-08
Time:                        19:38:51   Log-Likelihood:                -685.97
No. Observations:                 487   AIC:                             1382.
Df Residuals:                     482   BIC:                             1403.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   7.3362    



Based on this Adj R^2 value, my multi-linear model looking at Bar Ratings in Montreal conveys that popularity, the number of reviews, the price, the number of nearby bixi stations, and the busyness of those stations has a statistically significant relevance for 7% (0.071 R^2) of rating of a bar in Montreal. When popularity is considered, but what is interesting is that when rating is used to determine popularity, it has a p-value of 0.00 but now, in this case, the p-value for how popularity can reliably predict rating, it skyrockets to 0.811. This seems to show that online rating is a far better predictor of popularity than popularity is for a bar's rating. 
When I remove popularity, the adjusted R-squared value becomes 0.073. So a small improvement to the model.


From the bixi API, it is really interesting to see that the availability of affordable, nearby public-transit access is the most statistically significant element since it's p-value is 0.000. Interestingly, for rating, it also shows that number of reviews people left (p-value of 0.00) is also show statisticly significant when determining a bar's rating. That makes total sense because people leaving reviews is how a bar gets its rating.   think the intersting thing about the relationship between the number of reviews left and a bar's rating is that the coefficent of review_count is 0.0021. To me, this means that there is a lightly positive correlation between the number of reviews left and a bar's rating. That means that, as we might expect, the quality of the reviews is pretty close to 50-50 for good and bad reviews, but that generally most reviews are positive ones. 

Also, at least in this model, the price, with it's p-value of 0.043, is the weakest metric for determining a bar's rating. The interesting thing for future study would be a bigger sample of Yelp and Foursquare data because based on the top-10s of each website, it appeared as though Yelp's users were more concerned with price than the users of Foursquare.

The multivariate model also confirms the data of the simple linear regression which showed that num_stations_pinged is a statisticly significant metric when trying to predict a bar's online rating. Since people need affordable means of getting to and from a bar, and also since they cannot leagally drink and drive this makes sense. This data leads me to believe that there is an important possible exploreation for the business importance of having affordable transportation to and from bars and restaurants. 